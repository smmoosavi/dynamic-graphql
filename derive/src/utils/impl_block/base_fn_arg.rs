use darling::util::Ignored;
use syn::spanned::Spanned;

use crate::utils::impl_block::FromFnArg;
use crate::utils::impl_block::SelfArg;
use crate::utils::impl_block::TypedArg;
use crate::utils::with_context::SetContext;
use crate::utils::with_index::SetIndex;

#[derive(Debug, Clone)]
pub enum BaseFnArg {
    Receiver(SelfArg),
    Typed(TypedArg),
}

impl BaseFnArg {
    #[allow(dead_code)]
    pub fn get_attrs(arg: &syn::FnArg) -> &Vec<syn::Attribute> {
        match arg {
            syn::FnArg::Receiver(r) => &r.attrs,
            syn::FnArg::Typed(t) => &t.attrs,
        }
    }
    pub fn get_attrs_mut(arg: &mut syn::FnArg) -> &mut Vec<syn::Attribute> {
        match arg {
            syn::FnArg::Receiver(r) => &mut r.attrs,
            syn::FnArg::Typed(t) => &mut t.attrs,
        }
    }
}

impl BaseFnArg {
    pub fn span(&self) -> proc_macro2::Span {
        match self {
            Self::Receiver(r) => r.span,
            Self::Typed(t) => t.ident.span(),
        }
    }
}

impl SetIndex for BaseFnArg {
    fn with_index(self, _index: usize) -> Self {
        self
    }
}

impl SetContext for BaseFnArg {
    type Context = Ignored;

    fn set_context(&mut self, _: Self::Context) {}
}

impl FromFnArg for BaseFnArg {
    fn from_fn_arg(arg: &mut syn::FnArg) -> darling::Result<Self> {
        match arg {
            syn::FnArg::Receiver(receiver) => Ok(Self::Receiver(SelfArg {
                is_mut: receiver.mutability.is_some(),
                is_ref: receiver.reference.is_some(),
                span: receiver.self_token.span(),
            })),
            syn::FnArg::Typed(typed) => Ok({
                let ident = match *typed.pat {
                    syn::Pat::Ident(ref i) => i.ident.clone(),
                    _ => {
                        return Err(darling::Error::unsupported_shape("unnamed arguments")
                            .with_span(&typed.pat));
                    }
                };
                Self::Typed(TypedArg {
                    ident,
                    ty: typed.ty.as_ref().clone(),
                })
            }),
        }
    }
}
