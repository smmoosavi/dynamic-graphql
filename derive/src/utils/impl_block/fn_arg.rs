use crate::utils::error::GeneratorResult;
use crate::utils::with_index::SetIndex;
use darling::util::Ignored;
use syn::spanned::Spanned;

pub trait FromFnArg: Sized {
    fn from_fn_arg(arg: &mut syn::FnArg) -> GeneratorResult<Self>;
}

#[derive(Debug, Clone)]
pub struct SelfArg {
    pub is_mut: bool,
    pub is_ref: bool,
    pub span: proc_macro2::Span,
}

#[derive(Debug, Clone)]
pub struct TypedArg {
    pub ident: syn::Ident,
    pub ty: syn::Type,
}

#[derive(Debug, Clone)]
pub enum BaseFnArg {
    Receiver(SelfArg),
    Typed(TypedArg),
}

impl BaseFnArg {
    #[allow(dead_code)]
    pub fn get_attrs(arg: &syn::FnArg) -> &Vec<syn::Attribute> {
        match arg {
            syn::FnArg::Receiver(r) => &r.attrs,
            syn::FnArg::Typed(t) => &t.attrs,
        }
    }
    pub fn get_attrs_mut(arg: &mut syn::FnArg) -> &mut Vec<syn::Attribute> {
        match arg {
            syn::FnArg::Receiver(r) => &mut r.attrs,
            syn::FnArg::Typed(t) => &mut t.attrs,
        }
    }
}

impl Spanned for BaseFnArg {
    fn span(&self) -> proc_macro2::Span {
        match self {
            Self::Receiver(r) => r.span,
            Self::Typed(t) => t.ident.span(),
        }
    }
}

impl SetIndex for BaseFnArg {
    type Output = Self;
    fn with_index(self, _index: usize) -> Self {
        self
    }
}

impl FromFnArg for BaseFnArg {
    fn from_fn_arg(arg: &mut syn::FnArg) -> GeneratorResult<Self> {
        match arg {
            syn::FnArg::Receiver(receiver) => Ok(Self::Receiver(SelfArg {
                is_mut: receiver.mutability.is_some(),
                is_ref: receiver.reference.is_some(),
                span: receiver.span(),
            })),
            syn::FnArg::Typed(typed) => Ok({
                let ident = match *typed.pat {
                    syn::Pat::Ident(ref i) => i.ident.clone(),
                    _ => {
                        return Err(syn::Error::new(
                            typed.pat.span(),
                            "Only named arguments are supported",
                        )
                        .into());
                    }
                };
                Self::Typed(TypedArg {
                    ident,
                    ty: typed.ty.as_ref().clone(),
                })
            }),
        }
    }
}

impl FromFnArg for Ignored {
    fn from_fn_arg(_arg: &mut syn::FnArg) -> GeneratorResult<Self> {
        Ok(Ignored)
    }
}
