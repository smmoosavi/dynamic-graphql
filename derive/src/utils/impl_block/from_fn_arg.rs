use darling::util::Ignored;

pub trait FromFnArg: Sized {
    fn from_fn_arg(arg: &mut syn::FnArg) -> darling::Result<Self>;
}

#[derive(Debug, Clone)]
pub struct SelfArg {
    #[allow(dead_code)]
    pub is_mut: bool,
    #[allow(dead_code)]
    pub is_ref: bool,
    pub span: proc_macro2::Span,
}

#[derive(Debug, Clone)]
pub struct TypedArg {
    pub ident: syn::Ident,
    pub ty: syn::Type,
}

impl FromFnArg for Ignored {
    fn from_fn_arg(_arg: &mut syn::FnArg) -> darling::Result<Self> {
        Ok(Ignored)
    }
}
