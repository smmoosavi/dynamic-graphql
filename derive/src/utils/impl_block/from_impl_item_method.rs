use darling::util::Ignored;

pub trait FromImplItemFn: Sized {
    fn from_impl_item_method(impl_item_method: &mut syn::ImplItemFn) -> darling::Result<Self>;
}

impl FromImplItemFn for Ignored {
    fn from_impl_item_method(_method: &mut syn::ImplItemFn) -> darling::Result<Self> {
        Ok(Ignored)
    }
}
