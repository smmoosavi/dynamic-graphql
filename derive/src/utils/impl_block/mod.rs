pub use args::Args;
pub use base_fn_arg::BaseFnArg;
pub use base_item_impl::BaseItemImpl;
pub use base_item_trait::BaseItemTrait;
pub use base_method::BaseMethod;
pub use from_fn_arg::FromFnArg;
pub use from_fn_arg::SelfArg;
pub use from_fn_arg::TypedArg;
pub use from_impl_item_method::FromImplItemFn;
pub use from_item_impl::FromItemImpl;
pub use from_item_trait::FromItemTrait;
pub use from_signature::FromSignature;
pub use from_trait_item_method::FromTraitItemFn;
pub use methods::Methods;

mod base_item_impl;
mod from_item_impl;

mod base_fn_arg;
mod from_fn_arg;

mod from_impl_item_method;
mod from_trait_item_method;

mod base_item_trait;
mod from_item_trait;

mod base_method;
mod from_signature;

mod args;
mod methods;
