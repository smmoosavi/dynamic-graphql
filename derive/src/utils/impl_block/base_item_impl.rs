use darling::FromGenerics;

use crate::utils::impl_block::BaseMethod;
use crate::utils::impl_block::FromImplItemFn;
use crate::utils::impl_block::Methods;
use crate::utils::with_context::SetContext;
use crate::utils::with_index::SetIndex;
use crate::FromItemImpl;

#[derive(Debug, Clone)]
pub struct BaseItemImpl<Method = BaseMethod, Generics = ()> {
    pub trait_: Option<syn::Path>,
    pub ty: syn::Type,
    pub methods: Methods<Method>,
    pub generics: Generics,
    // todo generics, consts, types
}

impl<Method, Generics> FromItemImpl for BaseItemImpl<Method, Generics>
where
    Method: FromImplItemFn + SetIndex,
    Generics: FromGenerics,
{
    fn from_item_impl(item_impl: &mut syn::ItemImpl) -> darling::Result<Self> {
        Ok(Self {
            trait_: item_impl.trait_.as_ref().map(|t| t.1.clone()),
            ty: item_impl.self_ty.as_ref().clone(),
            generics: FromGenerics::from_generics(&item_impl.generics)?,
            methods: Methods::from_impl_item_methods(&mut item_impl.items.iter_mut())?,
        })
    }
}

impl<Method, Generics> SetContext for BaseItemImpl<Method, Generics>
where
    Method: FromImplItemFn + SetContext,
{
    type Context = Method::Context;

    fn set_context(&mut self, context: Self::Context) {
        self.methods.set_context(context);
    }
}
