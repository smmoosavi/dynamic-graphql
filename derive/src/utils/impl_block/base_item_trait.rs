use darling::FromGenerics;

use crate::utils::impl_block::from_item_trait::FromItemTrait;
use crate::utils::impl_block::from_trait_item_method::FromTraitItemFn;
use crate::utils::impl_block::BaseMethod;
use crate::utils::impl_block::Methods;
use crate::utils::with_context::SetContext;
use crate::utils::with_index::SetIndex;

#[derive(Clone, Debug)]
pub struct BaseItemTrait<Method = BaseMethod, Generics = ()> {
    pub ident: syn::Ident,
    pub generics: Generics,
    pub methods: Methods<Method>,
    // todo consts, types, super_traits
}

impl<Method, Generics> FromItemTrait for BaseItemTrait<Method, Generics>
where
    Method: FromTraitItemFn + SetIndex,
    Generics: FromGenerics,
{
    fn from_item_trait(item_trait: &mut syn::ItemTrait) -> darling::Result<Self> {
        Ok(Self {
            ident: item_trait.ident.clone(),
            generics: FromGenerics::from_generics(&item_trait.generics)?,
            methods: Methods::from_trait_item_methods(&mut item_trait.items.iter_mut())?,
        })
    }
}

impl<Method, Generics> SetContext for BaseItemTrait<Method, Generics>
where
    Method: FromTraitItemFn + SetContext,
{
    type Context = Method::Context;

    fn set_context(&mut self, context: Self::Context) {
        self.methods.set_context(context);
    }
}
