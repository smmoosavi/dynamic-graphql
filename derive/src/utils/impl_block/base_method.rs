use crate::utils::impl_block::Args;
use crate::utils::impl_block::BaseFnArg;
use crate::utils::impl_block::FromFnArg;
use crate::utils::impl_block::FromImplItemFn;
use crate::utils::impl_block::FromSignature;
use crate::utils::impl_block::FromTraitItemFn;
use crate::utils::with_context::SetContext;
use crate::utils::with_index::SetIndex;

#[derive(Debug, Clone)]
pub struct BaseMethod<MethodArg = BaseFnArg> {
    pub asyncness: bool,
    pub ident: syn::Ident,
    pub args: Args<MethodArg>,
    pub output_type: Option<syn::Type>,
}

impl<MethodArg: FromFnArg + SetIndex> FromSignature for BaseMethod<MethodArg> {
    fn from_signature(sig: &mut syn::Signature) -> darling::Result<Self> {
        Ok(BaseMethod {
            asyncness: sig.asyncness.is_some(),
            ident: sig.ident.clone(),
            args: Args::from_fn_args(&mut sig.inputs.iter_mut())?,
            output_type: match &sig.output {
                syn::ReturnType::Default => None,
                syn::ReturnType::Type(_, ty) => Some(ty.as_ref().clone()),
            },
        })
    }
}

impl<MethodArg: FromFnArg + SetIndex> FromImplItemFn for BaseMethod<MethodArg> {
    fn from_impl_item_method(impl_item_method: &mut syn::ImplItemFn) -> darling::Result<Self> {
        Self::from_signature(&mut impl_item_method.sig)
    }
}

impl<MethodArg: FromFnArg + SetIndex> FromTraitItemFn for BaseMethod<MethodArg> {
    fn from_trait_item_method(trait_item_method: &mut syn::TraitItemFn) -> darling::Result<Self> {
        Self::from_signature(&mut trait_item_method.sig)
    }
}

impl<A> SetIndex for BaseMethod<A> {
    fn with_index(self, _index: usize) -> Self {
        self
    }
}

impl<MethodArg> SetContext for BaseMethod<MethodArg> {
    type Context = ();

    fn set_context(&mut self, _context: Self::Context) {}
}
