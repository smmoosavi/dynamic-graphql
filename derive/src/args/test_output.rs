use super::*;
use crate::utils::impl_block::FromItemImpl;
use crate::FromItemTrait;
use darling::FromDeriveInput;
use prettier_please::unparse;
use unindent::Unindent;

fn pretty_expand(tokens: impl quote::ToTokens) -> String {
    let tokens = tokens.into_token_stream();
    let file: syn::File = syn::parse2(tokens).unwrap();
    unparse(&file)
}

fn derive<D: FromDeriveInput>(input: impl AsRef<str>) -> D {
    let input = input.as_ref();
    let file: syn::DeriveInput = syn::parse_str(input).unwrap();
    D::from_derive_input(&file).unwrap()
}

fn pretty_derive<D: FromDeriveInput + quote::ToTokens>(input: impl AsRef<str>) -> String {
    let input = input.as_ref().unindent();
    let derived = derive::<D>(&input);
    let pretty = pretty_expand(&derived);
    format!("{}\n{}", input, pretty)
}

fn expand_item_impl<D: FromItemImpl>(input: impl AsRef<str>) -> D {
    let input = input.as_ref();
    let mut file: syn::ItemImpl = syn::parse_str(input).unwrap();
    D::from_item_impl(&mut file).unwrap()
}

fn pretty_expand_item_impl<D: FromItemImpl + quote::ToTokens>(input: impl AsRef<str>) -> String {
    let expanded = expand_item_impl::<D>(&input);
    let pretty = pretty_expand(&expanded);
    format!("{}\n{}", input.as_ref().unindent(), pretty)
}

fn expand_item_trait<D: FromItemTrait>(input: impl AsRef<str>) -> D {
    let input = input.as_ref();
    let mut file: syn::ItemTrait = syn::parse_str(input).unwrap();
    D::from_item_trait(&mut file).unwrap()
}

fn pretty_expand_item_trait<D: FromItemTrait + quote::ToTokens>(input: impl AsRef<str>) -> String {
    let expanded = expand_item_trait::<D>(&input);
    let pretty = pretty_expand(&expanded);
    format!("{}\n{}", input.as_ref().unindent(), pretty)
}

fn md(outputs: &[&str]) -> String {
    format!("```rust\n{}\n```", outputs.join("\n\n"))
}

#[test]
fn test_app() {
    let input = r#"
        struct App(QueryRoot, PostApp);
    "#;

    let pretty = pretty_derive::<App>(input);
    let output = md(&[&pretty]);
    insta::assert_snapshot!(output);
}

#[test]
fn test_expand_object() {
    let input1 = r#"
        struct ExampleQuery<'a>(&'a Query);
    "#;

    let input2 = r#"
    impl ExampleQuery<'_> {
        fn the_example(&self) -> Example {
            Example {
                field: "field".to_string(),
            }
        }
    }
    "#;

    let pretty1 = pretty_derive::<ExpandObject>(input1);
    let pretty2 = pretty_expand_item_impl::<ExpandObjectFields>(input2);
    let output = md(&[&pretty1, &pretty2]);
    insta::assert_snapshot!(output);
}

#[test]
fn test_enum() {
    let input = r#"
        enum Example {
            Foo,
            Bar,
        }
    "#;
    let pretty = pretty_derive::<Enum>(input);
    let output = md(&[&pretty]);
    insta::assert_snapshot!(output);
}

#[test]
fn test_input_object() {
    let input = r#"
        struct ExampleInput {
            pub string: String,
        }
    "#;

    let pretty = pretty_derive::<InputObject>(input);
    let output = md(&[&pretty]);
    insta::assert_snapshot!(output);
}

#[test]
fn test_interface() {
    let input = r#"
        trait Node {
            fn id(&self) -> String;
        }

    "#;

    let pretty = pretty_expand_item_trait::<Interface>(input);
    let output = md(&[&pretty]);
    insta::assert_snapshot!(output);
}

#[test]
fn test_mutation() {
    let input1 = r#"
        struct MutationRoot;
    "#;

    let input2 = r#"
        struct MyMutation(MutationRoot);
    "#;
    let input3 = r#"
        impl MyMutation {
            fn the_example() -> String {
                "field".to_string()
            }
        }
    "#;

    let pretty1 = pretty_derive::<MutationRoot>(input1);
    let pretty2 = pretty_derive::<Mutation>(input2);
    let pretty3 = pretty_expand_item_impl::<MutationFields>(input3);
    let output = md(&[&pretty1, &pretty2, &pretty3]);
    insta::assert_snapshot!(output);
}

#[test]
fn test_resoled_object() {
    let input1 = r#"
        struct Example {
            pub field: String,
        }
    "#;

    let input2 = r#"
        impl Example {
            fn field(&self) -> &str {
                &self.field
            }
        }
    "#;

    let pretty1 = pretty_derive::<ResolvedObject>(input1);
    let pretty2 = pretty_expand_item_impl::<ResolvedObjectFields>(input2);
    let output = md(&[&pretty1, &pretty2]);
    insta::assert_snapshot!(output);
}

#[test]
fn test_scalar() {
    let input = r#"
        struct Example;
    "#;

    let pretty = pretty_derive::<Scalar>(input);
    let output = md(&[&pretty]);
    insta::assert_snapshot!(output);
}

#[test]
fn test_simple_object() {
    let input = r#"
        struct Example {
            pub field: String,
        }
    "#;

    let pretty = pretty_derive::<SimpleObject>(input);
    let output = md(&[&pretty]);
    insta::assert_snapshot!(output);
}

#[test]
fn test_union() {
    let input = r#"
        enum Animal {
            Dog(Dog),
            Cat(Cat),
        }
    "#;

    let pretty = pretty_derive::<Union>(input);
    let output = md(&[&pretty]);
    insta::assert_snapshot!(output);
}
