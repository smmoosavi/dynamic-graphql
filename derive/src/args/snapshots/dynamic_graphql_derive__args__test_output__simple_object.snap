---
source: derive/src/args/test_output.rs
expression: output
---
```rust
struct Example {
    pub field: String,
}

impl dynamic_graphql::internal::ParentType for Example {
    type Type = Example;
}
impl dynamic_graphql::internal::TypeName for Example {
    fn get_type_name() -> std::borrow::Cow<'static, str> {
        "Example".into()
    }
}
impl dynamic_graphql::internal::OutputTypeName for Example {}
impl dynamic_graphql::internal::Object for Example {}
impl<
    '__dynamic_graphql_lifetime,
> dynamic_graphql::internal::ResolveOwned<'__dynamic_graphql_lifetime> for Example {
    fn resolve_owned(
        self,
        _ctx: &dynamic_graphql::Context,
    ) -> dynamic_graphql::Result<
        Option<dynamic_graphql::FieldValue<'__dynamic_graphql_lifetime>>,
    > {
        Ok(Some(dynamic_graphql::FieldValue::owned_any(self)))
    }
}
impl<
    '__dynamic_graphql_lifetime,
> dynamic_graphql::internal::ResolveRef<'__dynamic_graphql_lifetime> for Example {
    fn resolve_ref(
        &'__dynamic_graphql_lifetime self,
        _ctx: &dynamic_graphql::Context,
    ) -> dynamic_graphql::Result<
        Option<dynamic_graphql::FieldValue<'__dynamic_graphql_lifetime>>,
    > {
        Ok(Some(dynamic_graphql::FieldValue::borrowed_any(self)))
    }
}
impl Example {
    fn __resolve_field(&self) -> &String {
        &self.field
    }
}
impl dynamic_graphql::internal::Register for Example {
    fn register(
        registry: dynamic_graphql::internal::Registry,
    ) -> dynamic_graphql::internal::Registry {
        let registry = registry.register::<String>();
        let object = dynamic_graphql::dynamic::Object::new(
            <Self as dynamic_graphql::internal::Object>::get_object_type_name().as_ref(),
        );
        let field = dynamic_graphql::dynamic::Field::new(
            "field",
            <String as dynamic_graphql::internal::GetOutputTypeRef>::get_output_type_ref(),
            |ctx| {
                dynamic_graphql::dynamic::FieldFuture::new(async move {
                    let parent = ctx.parent_value.try_downcast_ref::<Self>()?;
                    let value = Self::__resolve_field(parent);
                    dynamic_graphql::internal::Resolve::resolve(value, &ctx)
                })
            },
        );
        let object = object.field(field);
        registry.register_type(object)
    }
}

```
